### PR TITLE
Expand codegen to allow specify guards and async

### PR DESCRIPTION
**Pull request by DoumanAsh**
Route's attributes:

- `"path"` - Raw literal string with path for which to register handler
- `async` - Attribute to indicate that registered function is asynchronous.
- `guard="function_name"` - Registers function as guard using `actix_web::guard::fn_guard`

If resource type is not specified, attempts to infer it based on return type (if it fails, assume regular function)

### CODE REVIEW

File: actix-web-codegen/src/lib.rs [hunk 1]
```diff
@@ -1,118 +1,171 @@
...
-/// #[get("path")] attribute
+/// Creates route handler without attached method guard.
+///
+/// Syntax: `#[handler("path"[, attributes])]`
+///
+/// ## Attributes:
+///
+/// - `"path"` - Raw literal string with path for which to register handler. Mandatory.
+/// - `async` - Attribute to indicate that registered function is asynchronous.
+/// - `sync` - Attribute to indicate that registered as regular function.
```
Comment (fafhrd91) [inline actix-web-codegen/src/lib.rs]: cant you auto detect async/sync functions by `impl Future`?

File: actix-web-codegen/src/lib.rs [hunk 2]
```diff
@@ -1,118 +1,171 @@
...
+/// - `guard="function_name"` - Registers function as guard using `actix_web::guard::fn_guard`
```
Comment (fafhrd91) [inline actix-web-codegen/src/lib.rs:52]: we probably should add some predefined guards, something like `content_type = "text/plain"`

**Comment by fafhrd91** [comment]
regarding scope. first of all syntax is ugly.
much better to have resource registration, i.e:

```rust
#[service]
struct MyResource {
   ...
}

impl MyResource {
   #[get]
   fn index(&self) -> impl Responder {
       ...
   }

   #[post]
   fn update(&self) -> impl Responder {
       ...
   }

   #[get("/sub-path/index.html")]
   fn sub_index(&self) -> impl Responder {
       ...
   }
}
```

then you can register it with normal `App::new().service(MyResource)`
or within scope `App::new().service(web::scope("/prefix").service(MyResource)`

also what would be also good is async service initialization with custom parameters.
so developer could encapsulate resource initialization within service, something like redis or db connection

File: actix-web-codegen/src/lib.rs (hunk 1 above)
Comment (DoumanAsh) [reply actix-web-codegen/src/lib.rs]: I do auto detect it, if you take a look at tests, some of them don't mark as async/sync
These markers are mostly for cases when user want to override it

**Comment by DoumanAsh** [comment]
@fafhrd91 regarding that, I thought about it. 

But the problem will be is that we will not be able to properly implement it as we already have macros for routes that generate own structs inside these `impl` so it is going to break stuff.
Unless it is possible to use `get` like macros in both outside of `Service` and within, I'm not sure about it.
Also `Service` derive will need to be over `impl MyService` rather than over struct definition itself, as we'll need to analyze methods
Anyway I'll think about it

File: actix-web-codegen/src/lib.rs (hunk 1 above)
Comment (fafhrd91) [reply actix-web-codegen/src/lib.rs]: If you auto detect then remove attributes.

Comment (DoumanAsh) [reply actix-web-codegen/src/lib.rs]: You cannot auto-detect everything, user might have `MyBoxedAlias` type and there is no way to reliable detect whether it is future or not

Comment (DoumanAsh) [reply actix-web-codegen/src/lib.rs]: I'll remove `sync` since it is default

File: actix-web-codegen/src/lib.rs (hunk 2 above)
Comment (DoumanAsh) [reply actix-web-codegen/src/lib.rs:52]: Do you mean it also as having these guards in `actix_web::guard`?

File: actix-web-codegen/src/lib.rs (hunk 1 above)
Comment (fafhrd91) [reply actix-web-codegen/src/lib.rs]: we dont need to support 100% of use cases,  80% is more than enough, but those 80% should be easy to use.

File: actix-web-codegen/src/lib.rs (hunk 2 above)
Comment (fafhrd91) [reply actix-web-codegen/src/lib.rs:52]: no, just for proc macros.

**Comment by DoumanAsh** [comment]
Remove `scope` macro

File: actix-web-codegen/src/lib.rs (hunk 1 above)
Comment (DoumanAsh) [reply actix-web-codegen/src/lib.rs]: Remove `sync` but `async` remains. Though it is optional for use

**Comment by DoumanAsh** [comment]
>also what would be also good is async service initialization with custom parameters.
so developer could encapsulate resource initialization within service, something like redis or db connection

Does HttpServiceFactory allows for async initialization?

**Comment by fafhrd91** [comment]
It does

**Comment by DoumanAsh** [comment]
@fafhrd91 I have some troubles right now to come up with proper macro for service you want.
I'm a bit unsure how to handle routing there.
Since if we're going to make completely new service we cannot really use `Route` or `Scope`.
Which would make me write lots of routing related code in generated code

But I give another go to a scope macro and came up with something like that:

```rust
#[derive(Scope)]
pub struct MyScope {
    #[service]
    api: Api
    #[guard]
    guards: actix_web::guard::AllGuard
}

[#scope]
impl MyScope {
    [#get("/")]
    fn index() -> impl Future<Item=HttpResponse, Error=actix_web::Error> {
        future::ok(HttpResponse::Ok().finish())
     }
}
```

**Comment by fafhrd91** [comment]
Idea is interesting, but I don’t see much value in macros for scope. I’d rather have struct as a service impl. Yes, that would require to create completely new service and router. But if you want to give users to use &mut self in theirs method then you have to build new routing, existing functionality does not provide any of it

**Comment by DoumanAsh** [comment]
@fafhrd91 Scope macro is mostly to segregate parts of your routes into separate components instead of writing boilerplate yourself

**Comment by fafhrd91** [comment]
It is worth of solving this by macro. Something like mass service registration is enough, ie .services([route1,route2,.. ])

I am fine with simple macros that we have now.

**Comment by DoumanAsh** [comment]
@fafhrd91 Well, if we don't need scope macros then let's merge it as it is.
I think macros for service implementation might be overly complex to include in this PR

**Comment by fafhrd91** [comment]
Ok, makes sense. I’ll review later today

**Comment by DoumanAsh** [comment]
Ok, I'll publish my other macros as library on crates.io later on.
Then i'll think about services

File: actix-web-codegen/src/lib.rs [hunk 3]
```diff
@@ -1,118 +1,109 @@
 #![recursion_limit = "512"]
+//! Actix-web codegen module
+//!
+//! Generators for routes and scopes
+//!
+//! ## Route
+//!
+//! Macros:
+//!
+//! - [handler](attr.handler.html)
```
Comment (fafhrd91) [inline actix-web-codegen/src/lib.rs]: remove this

File: actix-web-codegen/src/lib.rs [hunk 4]
```diff
@@ -1,118 +1,109 @@
...
+//! - [get](attr.get.html)
+//! - [post](attr.post.html)
+//! - [put](attr.put.html)
+//! - [delete](attr.delete.html)
+//!
+//! ### Attributes:
+//!
+//! - `"path"` - Raw literal string with path for which to register handle. Mandatory.
+//! - `async` - Attribute to indicate that registered function is asynchronous.
```
Comment (fafhrd91) [inline actix-web-codegen/src/lib.rs]: remove this too

**Review by fafhrd91 — CHANGES_REQUESTED** [review]

File: actix-web-codegen/src/lib.rs (hunk 3 above)
Comment (DoumanAsh) [reply actix-web-codegen/src/lib.rs]: You mean to remove it from API?

File: actix-web-codegen/src/lib.rs (hunk 4 above)
Comment (DoumanAsh) [reply actix-web-codegen/src/lib.rs]: Ehm... I think it might be useful to specify when you cannot auto-detect asyncness... Are you sure?

Comment (fafhrd91) [reply actix-web-codegen/src/lib.rs]: Yes

File: actix-web-codegen/src/lib.rs (hunk 3 above)
Comment (fafhrd91) [reply actix-web-codegen/src/lib.rs]: Yes

File: actix-web-codegen/src/lib.rs (hunk 4 above)
Comment (DoumanAsh) [reply actix-web-codegen/src/lib.rs]: Done

File: actix-web-codegen/src/lib.rs (hunk 3 above)
Comment (DoumanAsh) [reply actix-web-codegen/src/lib.rs]: Done

**Comment by fafhrd91** [comment]
@DoumanAsh can we make something like:

```rust
#[get("/")]
#[guard="function"]
fn handler(...) -> {
   ...
}
```

**Comment by DoumanAsh** [comment]
I think it might be possible... but aside from flooding namespace, I'm not sure about order of evaluation for macros.
I suspect it would first come, first to be expand.

I'd say we should avoid creating too many attributes that are not part of Derive, but if you want I can try to check it out

**Comment by DoumanAsh** [comment]
@fafhrd91 I checked whether order of attributes matter.
If `#[get]` comes after `#[guard]` we will not be able to see `#[guard]` from within `#[get]`
I guess it could be possible to work it around, but that going to be a bit complicated if we're going to have multiple attributes like `#[guard]`

File: actix-web-codegen/src/route.rs [hunk 5]
```diff
@@ -0,0 +1,165 @@
+extern crate proc_macro;
+
+use std::fmt;
+
+use proc_macro::TokenStream;
+use quote::{quote};
+
+enum ResourceType {
```
Comment (fafhrd91) [inline actix-web-codegen/src/route.rs:8]: i dont think we need this

File: actix-web-codegen/src/route.rs [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        match self {
+            &ResourceType::Async => write!(f, "to_async"),
+            &ResourceType::Sync => write!(f, "to"),
+        }
+    }
+}
+
+#[derive(PartialEq)]
+pub enum GuardType {
+    None,
```
Comment (fafhrd91) [inline actix-web-codegen/src/route.rs]: do we use this?

File: actix-web-codegen/Cargo.toml [hunk 7]
```diff
@@ -13,9 +13,11 @@ proc-macro = true
 
 [dependencies]
 quote = "0.6"
+proc-macro2 = "0.4"
```
Comment (fafhrd91) [inline actix-web-codegen/Cargo.toml]: do you use this?

File: actix-web-codegen/src/route.rs (hunk 5 above)
Comment (DoumanAsh) [reply actix-web-codegen/src/route.rs:8]: I could use bool, but own enum makes it easier to just impl `Display` for it and use it in formatter

File: actix-web-codegen/src/route.rs (hunk 6 above)
Comment (DoumanAsh) [reply actix-web-codegen/src/route.rs]: Nope, removing it

File: actix-web-codegen/Cargo.toml (hunk 7 above)
Comment (DoumanAsh) [reply actix-web-codegen/Cargo.toml]: Nope, removing it

File: actix-web-codegen/src/route.rs (hunk 5 above)
Comment (fafhrd91) [reply actix-web-codegen/src/route.rs:8]: Right. It is ok. I misread it

**Review by fafhrd91 — APPROVED** [review]

**Pull request merged**